[WIP] PCA n_components='mle' instability. Issue 4441

lbillingham opened the pull request:
I'm only getting 
single lines in the compare view now for 

```
https://github.com/scikit-learn/scikit-learn/compare/master...lbillingham:issue_4441 
```

hopefully line endings are now actually posix-y.
Closes #4441.

Commit:
Partially fixes #4441. Protection from log(0) and spectrum with a tail -> 0.0 when using `pca(n_components='mle')`

lbillingham committed:
Partially fixes #4441. Protection from log(0) and spectrum with a tail -> 0.0 when using

Commit:
Partially fixes #4441.  Protection from log(0), tiny eigenvalues and spectrum with a tail -> 0.0 when using in PCA('mle')
 I've run into log(0) errors on data from the wild.
I haven't been able to construct a synthetic pathological X. Since _assess_dimension_ and _infer_dimension_ are imported in the tests I just test using a pathological spectrum.

Commit:
hopefully fixing line endings

amueller commented:
looks much better. Btw, you can force-push in old branches to not open new PRs. But you can also just close the other one.

amueller commented:
I would be great if in addition we had a test with a full dataset. If you have a perfectly correlated one, shouldn't that give you zeros? I think ``make_classification`` or ``make_low_rank_matrix`` should give you one of these.

lbillingham commented:
I'll take a look at better testing. Will prob not get to it for a couple of days though. 
Thanks for info about pushing to update an open PR

Sent from my iFern

> On 5 Jun 2015, at 21:14, Andreas Mueller <notifications@github.com> wrote:
> 
> I would be great if in addition we had a test with a full dataset. If you have a perfectly correlated one, shouldn't that give you zeros? I think make_classification or make_low_rank_matrix should give you one of these.
> 
> —
> Reply to this email directly or view it on GitHub.
>

vene commented on sklearn/decomposition/pca.py:
```diff
@@ -23,7 +24,7 @@
 
 def _assess_dimension_(spectrum, rank, n_samples, n_features):
-    """Compute the likelihood of a rank ``rank`` dataset
+    """Compute the likelihood of a rank ``rank`` dataset.
```
Compute the likelihood that dataset has the given rank?

vene commented on sklearn/decomposition/pca.py:
```diff
@@ -23,7 +24,7 @@
+    """Compute the likelihood of a rank ``rank`` dataset.
 
     The dataset is assumed to be embedded in gaussian noise of shape(n,
     dimf) having spectrum ``spectrum``.
```
does this mean shape `(n_samples, n_features)`?

vene commented on sklearn/decomposition/pca.py:
```diff
@@ -23,7 +24,7 @@
 
 
 def _assess_dimension_(spectrum, rank, n_samples, n_features):
```
why is this function both prefixed *and* suffixed by underscores? Shouldn't it be just the first?

amueller replied:
jupp. but that is not really part of the pr.

vene commented on sklearn/decomposition/tests/test_pca.py:
```diff
@@ -263,6 +263,40 @@ def test_infer_dim_3():
+    n_samples = 10
+    n_features = 5
+    ret = _infer_dimension_(spectrum, n_samples, n_features)
+    assert_equal(ret, 0)
```
Wait, there's something I don't get here.

Is the spectrum here supposed to be rank-1 or rank-0?

amueller replied:
I agree. it should be 1, right?

vene replied:
`_infer_dimension` takes the argmax. This might be an off-by-one indexing problem. If the second likelihood is the highest, the argmax returns 1 while `n_components` should be 2.  

Either way this test would be much clearer with the spectrum [1, 1e-30, 1e-30, 1e-30], or would that not work?

vene commented on sklearn/decomposition/pca.py:
```diff
@@ -68,6 +69,9 @@ def _assess_dimension_(spectrum, rank, n_samples, n_features):
         v = np.sum(spectrum[rank:]) / (n_features - rank)
         pv = -np.log(v) * n_samples * (n_features - rank) / 2.
 
+    if sys.float_info.epsilon > abs(v):
```
Is it safe to use this, should we use [`np.finfo`](http://docs.scipy.org/doc/numpy/reference/generated/numpy.finfo.html) instead, or should we just use a `tol` parameter or maybe a `rcond` as in [`scipy.pinv`](http://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.pinv.html)?

vene replied:
`sys.float_info.epsilon` means the difference between `1` and the smallest representable float greater than `1`. On my machine it's about 1e-16, and on 32-bit machines it should be about 1e-8. `np.finfo(np.float).tiny`, the smallest positive float, is much smaller. 

`pinv`'s `rcond` is 1e-15 by default:
```
    rcond : float
      Cutoff for small singular values.
      Singular values smaller (in modulus) than
      `rcond` * largest_singular_value (again, in modulus)
      are set to zero.
```

lbillingham committed:
sanitized test , explicitally added an  as an argument for ignoring small values rather than rely on , fixed function name underscoring

lbillingham committed:
Merge branch 'issue_4441' of https://github.com/lbillingham/scikit-learn into issue_4441

Conflicts:
	sklearn/decomposition/pca.py

lbillingham commented:
I've tried making a full test using 

```python
X, _ = sklearn.datasets.make_classification(n_repeated=17, n_informative=1, n_clusters_per_class=1, n_classes=1)
```
and
 
```python
X = sklearn.datasets..make_low_rank_matrix(effective_rank=2, tail_strength=1.0)
```

which, I think, are about as likely to cause an error as I can make them.

We do run into `v = 0 ` in `v = np.sum(spectrum[rank:]) / (n_features - rank)` but I have not been able to force a situation where we get a `log(0)`: . `spectrum_[rank:n_features] = v` and `for i in range(rank)` always conspire to avoid the `0`. 

If there is an off-by-one error and we should be looping to `range(rank + 1)` then the datasets above would cause `log(0)` errors.

I've changed the test `spectrum` to be as suggested and removed the trailing `_` from the function names.

I also noticed that 
```python
pv = -np.log(v) * n_samples * (n_features - rank) / 2.
```
was susceptible to `log(0)`, so I've moved the test of `if rcond > abs(v):` before it.

lbillingham commented:
I didn't mean to close this, sorry

lbillingham commented:
sorry, looks like I something wrong

lbillingham committed:
updating changes that hopefully fix the broken build. My repo was a mess.

amueller commented:
can you please add the test with a synthetic X and fixed random state?

lbillingham commented:
Sure: I'll get to it in ~12 hours or so. I don't think I can find a way to exercise the  protective code paths I created using the output of `make_low_rank_matrix` or `make_classification`.

amueller commented:
hum ok then. I'll try to have a look later, but maybe someone else finds time?

vene commented on sklearn/decomposition/pca.py:
```diff
@@ -75,6 +80,8 @@ def _assess_dimension_(spectrum, rank, n_samples, n_features):
     spectrum_ = spectrum.copy()
     spectrum_[rank:n_features] = v
     for i in range(rank):
+        if spectrum_[i] < rcond:
```
Can `spectrum_` values ever be negative here? If not, then the check above doesn't need the abs either, right?

lbillingham replied:
Spectrum comes from `S**2` with `_, S, _ = scipy.linalg.svd(X)` so I think it should non-negative.[we'll never see complex numbers will we?]

I'll have a go at constructing an `X` that gets trapped by the new code paths for a full test. But not been able to so far.

> On 9 Jun 2015, at 02:43, Vlad Niculae <notifications@github.com> wrote:
> 
> In sklearn/decomposition/pca.py:
> 
> > @@ -75,6 +80,8 @@ def _assess_dimension_(spectrum, rank, n_samples, n_features):
> >      spectrum_ = spectrum.copy()
> >      spectrum_[rank:n_features] = v
> >      for i in range(rank):
> > +        if spectrum_[i] < rcond:
> Can spectrum_ values ever be negative here? If not, then the check above doesn't need the abs either, right?
> 
> —
> Reply to this email directly or view it on GitHub.
>

vene replied:
> we'll never see complex numbers will we?
PCA could be one of the algorithms that might with complex numbers. But we don't support it AFAIK.

We can either remove the abs everywhere, or add it also inside the `v` summation. Couldn't say which is better.

lbillingham committed:
Merge branch 'master' of https://github.com/lbillingham/scikit-learn into issue_4441

lbillingham committed:
added a full test with `X` to exercise new  ` if rcond > v: return -np.inf`` codepath in `pca._assess_dimension`. Removed redundant `abs(v)` as `v` composed by summing squared real numbers.

lbillingham commented:
@vene you are right about the `abs()` being surplus to requirements. 
I've got an `X` that goes through the test on `rcond > v`, have not been able to make one for `spectrum_[i] < rcond`

lbillingham commented:
Anything else I need to do on this?

amueller commented:
Sorry, we have a lot of things to review and not enough people.

amueller commented on sklearn/decomposition/tests/test_pca.py:
```diff
@@ -260,7 +260,46 @@ def test_infer_dim_3():
+                                        n_redundant=1, n_clusters_per_class=1,
+                                        random_state=20150609)
+    pca = PCA(n_components='mle').fit(X)
+    assert_equal(pca.n_components_, 0)
```
That seems really odd to me. Shouldn't it be 1?

lbillingham replied:
Is this related to the off-by-one error that @alexis-mignon talked about in the original comment on 24 Mar?

lbillingham commented:
No worries, so long as I'm not causing a hold up

On 16 June 2015 at 16:18, Andreas Mueller <notifications@github.com> wrote:

> Sorry, we have a lot of things to review and not enough people.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4827#issuecomment-112467578>
> .
>

rth commented:
@lbillingham Could your please resolve conflicts or rebase? Thanks.

lbillingham commented:
Update on this, I currently only a work computer or a tablet that is not up to this kind of work.
I've asked if I can do FOSS work on my time but company hardware: hopefully that will be a yes. But it is taking some time.

rth commented:
@lbillingham If you don't have the availability to finish this PR, can it be continued by another contributor? Thanks.

yxliang01 commented:
Wow... This has been 4 years now... And, `mle` is still unusable to me... :joy:

jnothman commented:
It's yours to complete if you want, @yxliang01!

jnothman commented:
#10359 is the one to be completed, @yxliang01. Go on, take it over, address the last few comments.

jnothman commented:
I'm closing this one in preference for #10359.

yxliang01 commented:
@jnothman Unfortunately, I don't really understand MLE algorithm nor sklearn itself. So, at least for now, I am not suitable to do so...

jnothman commented:
Pity. It would be nice to see this completed and I think it is not far off

yxliang01 commented:
@jnothman Yeah...
